pkg/controller/queuejob/queuejob_controller_ex.go

```diff
@@ -757,12 +784,22 @@ func (qjm *XController) UpdateAgent() {
 }
 
 func (qjm *XController) UpdateQueueJobs() {
+	firstTime := metav1.NowMicro()
 	queueJobs, err := qjm.queueJobLister.AppWrappers("").List(labels.Everything())
 	if err != nil {
 		glog.Errorf("[UpdateQueueJobs] List of queueJobs err=%+v", err)
 		return
 	}
 	for _, newjob := range queueJobs {
+		// UpdateQueueJobs can be the first to see a new AppWrapper job, under heavy load
+		if newjob.Status.QueueJobState == "" {
+			newjob.Status.ControllerFirstTimestamp = firstTime
+			newjob.Status.SystemPriority = newjob.Spec.Priority
+			newjob.Status.QueueJobState = arbv1.QueueJobStateInit
+			glog.V(4).Infof("[UpdateQueueJobs] %s 0Delay=%.6f seconds CreationTimestamp=%s ControllerFirstTimestamp=%s",
+				newjob.Name, time.Now().Sub(newjob.Status.ControllerFirstTimestamp.Time).Seconds(), newjob.CreationTimestamp, newjob.Status.ControllerFirstTimestamp)
+		}
+		glog.V(10).Infof("[UpdateQueueJobs] %s: qjqueue=%t &qj=%p Version=%s Status=%+v", newjob.Name, qjm.qjqueue.IfExist(newjob), newjob, newjob.ResourceVersion, newjob.Status)
```

Comment:
I have a question regarding the following timing conditions that I consider could occur:

`UpdateQueueJobs()` thread gets an AW before the `worker()` thread 'completes' processing of the 'add'  the `newjob` will be processed as above in the UpdateQueueJobs() thread.  

My question is:

If the separate 'worker()' thread completes processing of the same AW and adds the new job in the `qjqueue` during the time between lines 803-805 (after the `if` statement and before the `enqueue`) does the current mcad logic behave as expected?  

I realize this is probably a condition that had already existed before the changes above but I wanted to find out your thoughts on the condition above since you are setting fields now before the `enqueue`.

Reply:
This is a good question regarding if an AW can be in qjqueue and eventQueue at the same time, and what to do.

We should design the system such that an AW should not be in qjqueue and eventQueue at the same time.  

1) The Add operation for both qjqueue and eventQueue guarantees there is only one AW (represented by namespace/name) in each queue.  

2) The worker thread pops the AW before adding it to qjqueue.  

3) The ScheduleNext thread removes AW from qjqueue after adding it to eventQueue.  

4) The Informer adds AW to eventQueue when it is seen first.  So the AW can not be in qjqueue.

5) The UpdateQueueJobs thread adds AW to eventQueue.  It should check to make sure the AW is not already in qjqueue.  If the AW is already in eventQueue, then it should leave the current copy of AW in the eventQueue, because the AW information obtained from the local store by the UpdateQueueJobs thread may not contain the most up-to-date information.

Item 5) above is addressed in the latest commit: 8d34048